File: MainProject/Scenes/HexPlanet/Util/TilePriorityQueue.cs
using System.Collections.Generic;
using ZeromaXsPlaygroundProject.Scenes.HexPlanet.Struct;

namespace ZeromaXsPlaygroundProject.Scenes.HexPlanet.Util;

public class TilePriorityQueue(TileSearchData[] data)
{
    private readonly List<int> _list = [];
    private int _minimum = int.MaxValue;

    public void Enqueue(int tileId)
    {
        var priority = data[tileId].SearchPriority;
        if (priority < _minimum)
            _minimum = priority;
        while (priority >= _list.Count)
            _list.Add(-1);
        data[tileId].NextWithSamePriority = _list[priority];
        _list[priority] = tileId;
    }
    
    public bool TryDequeue(out int tileId)
    {
        for (; _minimum < _list.Count; _minimum++)
        {
            tileId = _list[_minimum];
            if (tileId >= 0)
            {
                _list[_minimum] = data[tileId].NextWithSamePriority;
                return true;
            }
        }
        tileId = -1;
        return false;
    }

    public void Change(int tileId, int oldPriority)
    {
        var current = _list[oldPriority];
        var next = data[current].NextWithSamePriority;
        if (current == tileId)
            _list[oldPriority] = next;
        else
        {
            while (next != tileId)
            {
                current = next;
                next = data[current].NextWithSamePriority;
            }
            data[current].NextWithSamePriority = data[tileId].NextWithSamePriority;
        }
        Enqueue(tileId);
    }
    
    public void Clear()
    {
        _list.Clear();
        _minimum = int.MaxValue;
    }
}

File: TerraObserver/TO.Domains/Functions/PathFindings/TilePriorityQueue.fs
namespace TO.Domains.Functions.PathFindings

open System
open Friflo.Engine.ECS
open TO.Domains.Functions.HexSpheres.Components
open TO.Domains.Types.HexSpheres
open TO.Domains.Types.PathFindings

/// Copyright (C) 2025 Zhu Xiaohe(aka ZeromaXHe)
/// Author: Zhu XH (ZeromaXHe)
/// Date: 2025-06-29 19:52:29
module TilePriorityQueue =
    let enqueue (tileCountId: int) (this: TilePriorityQueue) =
        let priority = this.Data[tileCountId].SearchPriority

        if priority < this.Minimum then
            this.Minimum <- priority

        while priority >= this.List.Count do
            this.List.Add -1

        this.Data[tileCountId].NextWithSamePriority <- this.List[priority]
        this.List[priority] <- tileCountId

    let tryDequeue (tileId: TileId outref) (this: TilePriorityQueue) =
        let mutable result = false

        while this.Minimum < this.List.Count && not result do
            tileId <- this.List[this.Minimum]

            if tileId >= 0 then
                this.List[this.Minimum] <- this.Data[tileId].NextWithSamePriority
                result <- true
            else
                this.Minimum <- this.Minimum + 1

        if not result then
            tileId <- -1

        result

    let change (tileCountId: int) (oldPriority: int) (this: TilePriorityQueue) =
        let mutable current = this.List[oldPriority]
        let mutable next = this.Data[current].NextWithSamePriority

        if current = tileCountId then
            this.List[oldPriority] <- next
        else
            while next <> tileCountId do
                current <- next
                next <- this.Data[current].NextWithSamePriority

            this.Data[current].NextWithSamePriority <- this.Data[tileCountId].NextWithSamePriority

        enqueue tileCountId this

    let clear (this: TilePriorityQueue) =
        this.List.Clear()
        this.Minimum <- Int32.MaxValue
